doc/source/new-project-guide/level-3-tutorial.rst

```diff
@@ -48,22 +51,109 @@ When you ``cd`` into the new directory, you will see a folder structure as shown
+     import numpy as np
+
+
+     def dot_product(a, b):
+          """Calculate the dot product of two vectors."""
+          return np.dot(a, b)
+
+- ``proj_one/proj_one_code.py`` shows an example of how to import and use the ``dot_product()`` function from ``shared_functions.py``:
+
+.. code-block:: python
+
+     # proj_one_code.py
+     import shared_functions
+
+     a = [1, 2, 3]
+     b = [1, 2, 3]
+     result = shared_functions.dot_product(a, b)
+     print(result)
+
+- ``__init__.py`` files mark directories as Python modules. These are empty by default at this level.
+
+- ``requirements.txt`` lists your project's dependencies, which in this case includes ``numpy`` and ``pytest``:
+
+.. code-block:: python
+
+     # requirements.txt
+     numpy
+     pytest
+
+- The ``tests`` folder contains tests for your shared functions. For example, ``test_shared_functions.py`` includes tests for the ``dot_product`` function using ``pytest``. General naming convention for test files is ``test_<module_name>.py``. Similarly, test function names should be ``test_<function_name>()``. This is how ``pytest`` recognizes them as test functions:
+
+
+.. code-block:: python
+
+     # test_shared_functions.py
+     import numpy as np
+     import pytest
+     import shared_functions
+
+
+     def test_dot_product_2D_list():
+     a = [1, 2]
+     b = [3, 4]
+     expected = 11.0
+     actual = shared_functions.dot_product(a, b)
+     assert actual == expected
+
+
+     def test_dot_product_3D_list():
+     a = [1, 2, 3]
+     b = [4, 5, 6]
+     expected = 32.0
+     actual = shared_functions.dot_product(a, b)
+     assert actual == expected
 
-.. note::
 
-     ``shared_functions.py`` is an example module where you can define functions that are imported across the project folders ``proj-one`` and ``proj-two``. The ``__init__.py`` files are required and empty files that indicate to Python that the directories contain Python modules.
+     @pytest.mark.parametrize(
+     "a, b, expected",
+     [
+          # Test whether the dot product function works with 2D and 3D vectors
+          # C1: lists, expect correct float output
+          ([1, 2], [3, 4], 11.0),
+          ([1, 2, 3], [4, 5, 6], 32.0),
+          # C2: tuples, expect correct float output
+          ((1, 2), (3, 4), 11.0),
+          ((1, 2, 3), (4, 5, 6), 32.0),
+          # C3: numpy arrays, expect correct float output
+          (np.array([1, 2]), np.array([3, 4]), 11.0),
+          (np.array([1, 2, 3]), np.array([4, 5, 6]), 32.0),
+     ],
+     )
+     def test_dot_product(a, b, expected):
+     actual = shared_functions.dot_product(a, b)
+     assert actual == expected
 
-     ``requirements.txt`` is the file where you list Python dependencies. The ``README.md`` file is where you add notes for your project. The ``tests`` folder contains tests for your reusable code.
+- ``.pre-commit-config.yaml`` is a configuration file for pre-commit hooks. This file is described in further detail below.
```

Comment:
it is probably important to remind people that they must install pre-commit in their env (conda install pre-commit" but then run pre-commit for it to work "pre-commit install" or whatever (step by step instructions).  If this is described elsewhere, don't repeat it but refer to it here for completeness.